core/src/test/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerFacadeTest.java

```diff
@@ -0,0 +1,482 @@
+
+        ArgumentCaptor<MediaPlayer.OnCompletionListener> argumentCaptor = ArgumentCaptor.forClass(MediaPlayer.OnCompletionListener.class);
+        verify(mediaPlayer).setOnCompletionListener(argumentCaptor.capture());
+        argumentCaptor.getValue().onCompletion(mediaPlayer);
+
+        verify(completionListener).onCompletion(mediaPlayer);
+    }
+
+    @Ignore("Should we throw when a listener is not attached?")
+    @Test
+    public void givenNoBoundCompletionListener_andMediaPlayerIsPrepared_whenCompleted_thenDoesNotForwardCompleted() {
+
+    }
+
+    @Test
+    public void givenBoundErrorListener_andMediaPlayerIsPrepared_whenErroring_thenForwardsError() {
+        givenMediaPlayerIsPrepared();
+
+        ArgumentCaptor<MediaPlayer.OnErrorListener> argumentCaptor = ArgumentCaptor.forClass(MediaPlayer.OnErrorListener.class);
+        verify(mediaPlayer).setOnErrorListener(argumentCaptor.capture());
+        argumentCaptor.getValue().onError(mediaPlayer, ANY_ERROR_WHAT, ANY_ERROR_EXTRA);
+
+        verify(errorListener).onError(mediaPlayer, ANY_ERROR_WHAT, ANY_ERROR_EXTRA);
+    }
+
+    @Ignore("Should we throw when a listener is not attached?")
+    @Test
+    public void givenNoBoundErrorListener_andMediaPlayerIsPrepared_whenErroring_thenDoesNotForwardError() {
+
+    }
+
+    @Test
+    public void givenBoundBufferListener_andMediaPlayerIsPrepared_whenBuffering_thenBufferPercentageIsUpdated() {
+        givenMediaPlayerIsPrepared();
+
+        ArgumentCaptor<MediaPlayer.OnBufferingUpdateListener> argumentCaptor = ArgumentCaptor.forClass(MediaPlayer.OnBufferingUpdateListener.class);
+        verify(mediaPlayer).setOnBufferingUpdateListener(argumentCaptor.capture());
+        argumentCaptor.getValue().onBufferingUpdate(mediaPlayer, TEN_PERCENT);
+
+        int bufferPercentage = facade.getBufferPercentage();
+        assertThat(bufferPercentage).isEqualTo(TEN_PERCENT);
+    }
+
+    @Test
+    public void givenMediaPlayerIsPrepared_whenReleasing_thenReleasesMediaPlayer() {
+        givenMediaPlayerIsPrepared();
+
+        facade.release();
+
+        verify(mediaPlayer).reset();
+        verify(mediaPlayer).release();
+    }
+
+    @Ignore("We should forward to listeners / make logs testable / throw exception.")
+    @Test
+    public void givenNoBoundSurfaceHolderRequester_whenStarting_thenLogsNotAttachedWarning() {
+
+    }
+
+    @Test
+    public void givenMediaPlayerIsPrepared_whenStarting_thenSetsDisplay() {
+        givenMediaPlayerIsPrepared();
+        reset(mediaPlayer);
+
+        facade.start();
+
+        verify(mediaPlayer).setDisplay(surfaceHolder);
+    }
+
+    @Test
+    public void givenMediaPlayerIsNotPrepared_whenStarting_thenNeverSetsDisplay() {
+        facade.start();
+
+        verify(mediaPlayer, never()).setDisplay(surfaceHolder);
+    }
+
+    @Test
+    public void givenMediaPlayerIsPrepared_whenStarting_thenStartsMediaPlayer() {
+        givenMediaPlayerIsPrepared();
+        reset(mediaPlayer);
```

Comment:
There are two resets in this class the second of which is required.

`reset` is used to reset the `mediaPlayer` mock. We reset because in the `givenMediaPlayerIsPrepared` method we will have interactions with `MediaPlayer.setDisplay`, but we aren't interested in those, we are only interested in what occurs outside of it.